File: java/Exercises.java
import java.util.List;
import java.util.Objects;
import java.util.Map;
import java.util.HashMap;
import java.util.Optional;
import java.util.function.Predicate;
import java.util.ArrayList;  
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Exercises {
    static Map<Integer, Long> change(long amount) {
        if (amount < 0) {
            throw new IllegalArgumentException("Amount cannot be negative");
        }
        var counts = new HashMap<Integer, Long>();
        for (var denomination : List.of(25, 10, 5, 1)) {
            counts.put(denomination, amount / denomination);
            amount %= denomination;
        }
        return counts;
    }
    // Method to find the first string that matches a predicate and convert it to lowercase
    static Optional<String> firstThenLowerCase(List<String> list, Predicate<String> predicate) {
        if (list == null || list.isEmpty()) {
            return Optional.empty();
        }
        return list.stream()
                   .filter(predicate)
                   .findFirst()
                   .map(String::toLowerCase); // Fix: Add closing brace here
    }
    // SayBuilder class
    public static class SayBuilder {
        private final List<String> words;

        // Constructor for an empty phrase
        public SayBuilder() {
            this.words = new ArrayList<>();
        }

        // Constructor with initial words
        public SayBuilder(List<String> initialWords) {
            this.words = new ArrayList<>(initialWords);
        }
        // Method to add a word and return a new SayBuilder with updated words list
        public SayBuilder and(String word) {
            List<String> newWords = new ArrayList<>(words);
            newWords.add(word);
            return new SayBuilder(newWords);
        }
        // Method to generate the full phrase
        public String phrase() {
            return String.join(" ", words);
        }
    }
    // Method to create a SayBuilder with optional initial words
    public static SayBuilder say(String... words) {
        if (words == null || words.length == 0) {
            return new SayBuilder();
        } else {
            return new SayBuilder(List.of(words));
        }
    }
    // Method to count meaningful lines in a file
    static int meaningfulLineCount(String filePath) throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            return (int) reader.lines()
                               .map(String::trim) // Trim each line
                               .filter(line -> !line.isEmpty() && !line.startsWith("#")) // Check for meaningful lines
                               .count();
        }
    }
}

record Quaternion(double a, double b, double c, double d) {
    public static final Quaternion ZERO = new Quaternion(0, 0, 0, 0);
    public static final Quaternion I = new Quaternion(0, 1, 0, 0);
    public static final Quaternion J = new Quaternion(0, 0, 1, 0);
    public static final Quaternion K = new Quaternion(0, 0, 0, 1);
    private static final double EPSILON = 1e-9;
    // check for NaN
    public Quaternion {
        if (Double.isNaN(a) || Double.isNaN(b) || Double.isNaN(c) || Double.isNaN(d)) {
            throw new IllegalArgumentException("Coefficients cannot be NaN");
        }
    }
    // add this quaternion to another quaternion
    public Quaternion plus(Quaternion other) {
        return new Quaternion(a + other.a, b + other.b, c + other.c, d + other.d);
    }
    // * this quaternion by another quaternion
    public Quaternion times(Quaternion other) {
        return new Quaternion(
            a * other.a - b * other.b - c * other.c - d * other.d,
            a * other.b + b * other.a + c * other.d - d * other.c,
            a * other.c - b * other.d + c * other.a + d * other.b,
            a * other.d + b * other.c - c * other.b + d * other.a
        );
    }
    // return the conjugate of this quaternion
    public Quaternion conjugate() {
        return new Quaternion(a, -b, -c, -d);
    }
    // retune the coefficients
    public List<Double> coefficients() {
        return List.of(a, b, c, d);
    }
    // custom equality to handle floating-point precision
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Quaternion that = (Quaternion) o;
        return Math.abs(this.a - that.a) < EPSILON &&
               Math.abs(this.b - that.b) < EPSILON &&
               Math.abs(this.c - that.c) < EPSILON &&
               Math.abs(this.d - that.d) < EPSILON;
    }
    @Override
    public int hashCode() {
        return Objects.hash(a, b, c, d);
    }
    //  toString() 
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();

        if (a != 0) {
            sb.append(a);
        }
        appendTerm(sb, b, "i");
        appendTerm(sb, c, "j");
        appendTerm(sb, d, "k");
        return sb.length() == 0 ? "0" : sb.toString();
    }
    // helpr method to append terms and formatiing
    private void appendTerm(StringBuilder sb, double value, String letter) {
        if (value == 0) return;

        if (sb.length() > 0 && value > 0) {
            sb.append("+");
        }

        if (value == 1) {
            sb.append(letter);
        } else if (value == -1) {
            sb.append("-").append(letter);
        } else {
            sb.append(value).append(letter);
        }
    }
}

// sealed interface that allows only two implementing classes: 
sealed interface BinarySearchTree permits Empty, Node {
    int size(); // 
    boolean contains(String value); 
    BinarySearchTree insert(String value); 
}

// class representing an empty tree (base case)
final record Empty() implements BinarySearchTree {
    @Override
    public int size() {
        return 0; 
    }
    @Override
    public boolean contains(String value) {
        return false;
    }
    @Override
    public BinarySearchTree insert(String value) {
        // Inserting into an empty tree creates a new node with the value
        return new Node(value, this, this); 
    }
    @Override
    public String toString() {
        return "()"; 
    }
}
// class representing teh non-empty node i
final class Node implements BinarySearchTree {
    private final String value; 
    private final BinarySearchTree left; 
    private final BinarySearchTree right; 
    // constructor which initializex
    public Node(String value, BinarySearchTree left, BinarySearchTree right) {
        this.value = value;
        this.left = left;
        this.right = right;
    }
    @Override
    public int size() {
        return 1 + left.size() + right.size(); 
    }
    @Override
    public boolean contains(String value) {
        if (value.equals(this.value)) {
            return true;
        } else if (value.compareTo(this.value) < 0) {
            return left.contains(value); // check the left 
        } else {
            return right.contains(value); // check the right 
        }
    }
    @Override
    public BinarySearchTree insert(String value) {
        if (value.equals(this.value)) {
            return this; // if the value is already present, return the current node 
        } else if (value.compareTo(this.value) < 0) {
            return new Node(this.value, left.insert(value), right); // instyt into the left subtree
        } else {
            return new Node(this.value, left, right.insert(value)); // inesert into the right subtree
        }
    }
    @Override
    public String toString() {
        String leftStr = left.toString().equals("()") ? "" : left.toString();
        String rightStr = right.toString().equals("()") ? "" : right.toString();
        return "(" + leftStr + value + rightStr + ")"; 
    }
}

File: kotlin/Exercises.kt
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.io.File


fun change(amount: Long): Map<Int, Long> {
    require(amount >= 0) { "Amount cannot be negative" }
    
    val counts = mutableMapOf<Int, Long>()
    var remaining = amount
    for (denomination in listOf(25, 10, 5, 1)) {
        counts[denomination] = remaining / denomination
        remaining %= denomination
    }
    return counts
}

// Write your first then lower case function here
fun firstThenLowerCase(strings: List<String>, predicate: (String) -> Boolean): String? { 
    return string.firstOrNull(predicate)?.lowercase() // returns possible String, ?. operator which handles null cases
}

// Write your say function here
class say(val sentence: String = "") { // set original sentence to "" for empty string cases
    fun and(nextWord: String): say {
        return say("${this.sentence} $nextWord") 
    }

    val phrase: String 
        get() = this.sentence
}

// Write your meaningfulLineCount function here
fun meaningfulLineCount(filename: String): Long {
    return try {
        // open file with buffered reader
        BufferedReader(FileReader(File(filename))).use { reader ->
            var count: Long = 0 // counter to track lines
            reader.forEachLine { line ->
                // remove white space 
                val trimmedLine = line.trim()
                if (trimmedLine.isNotEmpty() && !trimmedLine.startsWith("#")) {
                    count++
                }
            }
            // return total count
            count
        }
    } catch (e: IOException) {
        // if not found or opened throw
        throw IOException("No such file: $filename", e)
    }
}

// Write your Quaternion data class here
data class Quaternion(val a: Double, val b: Double, val c: Double, val d: Double) { 

    companion object {
        val ZERO = Quaternion(0.0, 0.0, 0.0, 0.0)
        val I = Quaternion(0.0, 1.0, 0.0, 0.0)
        val J = Quaternion(0.0, 0.0, 1.0, 0.0)
        val K = Quaternion(0.0, 0.0, 0.0, 1.0)
    }
    operator fun plus(other: Quaternion): Quaternion { // using operator fun to overload + operator (addition)
        return Quaternion(a + other.a, b + other.b, c + other.c, d + other.d)
    }

    operator fun times(other: Quaternion): Quaternion { // using operator fun to overload * operator (multiplication)
        return Quaternion(
            a * other.a - b * other.b - c * other.c - d * other.d,
            a * other.b + b * other.a + c * other.d - d * other.c,
            a * other.c - b * other.d + c * other.a + d * other.b,
            a * other.d + b * other.c - c * other.b + d * other.a
        )
    }

    override fun toString(): String {
        val parts = mutableListOf<String>() // mutable so that the list's strings can be modified

        fun formatTerm(value: Double, letter: String): String {
            return when {
                value == 0.0 -> ""
                value == 1.0 -> letter
                value == -1.0 -> "-$letter"
                else -> "$value$letter"
            }
        }

        if (a != 0.0) parts.add("$a")
        if (b != 0.0) parts.add(formatTerm(b, "i"))
        if (c != 0.0) parts.add(formatTerm(c, "j"))
        if (d != 0.0) parts.add(formatTerm(d, "k"))

        if (parts.isEmpty()) return "0"

        var result = parts[0]
        for (part in parts.drop(1)) {
            result += if (part.startsWith("-")) part else "+$part"
        }

        return result

    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Quaternion) return false

        return this.a == other.a &&
               this.b == other.b &&
               this.c == other.c &&
               this.d == other.d
    }

    fun coefficients(): List<Double> = listOf(a, b, c, d)

    fun conjugate(): Quaternion = Quaternion(a, -b, -c, -d)
}



// Write your Binary Search Tree interface and implementing classes here
sealed interface BinarySearchTree {
    // method to insert a new value into the tree, returning a new instance 
    fun insert(value: String): BinarySearchTree
    // method to check if a value is contained in the tree
    fun contains(value: String): Boolean
    // methof to get the size of the tree (number of elements)
    fun size(): Int
    // method to get the size of the tree (number of elements)
    override fun to
    String(): String

//  representing an empty Binary Search Tree
    object Empty : BinarySearchTree {
        override fun insert(value: String): BinarySearchTree {
            return Node(value, Empty, Empty)
        }

        override fun contains(value: String): Boolean = false

        override fun size(): Int = 0

        override fun toString(): String = "()"
    }
    // data class representing a node in the Binary Search Tree

    data class Node(
        val value: String,
        val left: BinarySearchTree = Empty,
        val right: BinarySearchTree = Empty
    ) : BinarySearchTree {
          // insert a new value into the tree, maintaining binary search tree props

        override fun insert(newValue: String): BinarySearchTree {
            return when {
                newValue < value -> Node(value, left.insert(newValue), right)
                newValue > value -> Node(value, left, right.insert(newValue))
                else -> this // Prevent duplicates
            }
        }

        override fun contains(value: String): Boolean {
            return when {
                value < this.value -> left.contains(value)
                value > this.value -> right.contains(value)
                else -> true // Value found
            }
        }
       // calc the size of the tree, counting this node and its children

        override fun size(): Int = 1 + left.size() + right.size()
        // str representation of the tree in the expected format

        override fun toString(): String {
            val leftStr = left.toString().takeIf { it != "()" } ?: ""
            val rightStr = right.toString().takeIf { it != "()" } ?: ""
            return "$leftStr$value$rightStr".let { if (it.isNotEmpty()) "($it)" else it }
        }
    }
}
